lightning-invoice/src/payment.rs

```diff
@@ -244,137 +267,158 @@ where
 		if invoice.amount_milli_satoshis().is_some() {
 			Err(PaymentError::Invoice("amount unexpected"))
 		} else {
-			self.pay_invoice_internal(invoice, Some(amount_msats), 0)
+			self.pay_invoice_using_amount(invoice, Some(amount_msats))
 		}
 	}
 
-	fn pay_invoice_internal(
-		&self, invoice: &Invoice, amount_msats: Option<u64>, retry_count: usize
+	fn pay_invoice_using_amount(
+		&self, invoice: &Invoice, amount_msats: Option<u64>
 	) -> Result<PaymentId, PaymentError> {
 		debug_assert!(invoice.amount_milli_satoshis().is_some() ^ amount_msats.is_some());
+
 		let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
-		if invoice.is_expired() {
-			log_trace!(self.logger, "Invoice expired prior to first send for payment {}", log_bytes!(payment_hash.0));
+		let mut payment_cache = self.payment_cache.lock().unwrap();
+		let mut entry = match payment_cache.entry(payment_hash) {
+			hash_map::Entry::Occupied(_) => return Err(PaymentError::Invoice("payment pending")),
+			hash_map::Entry::Vacant(_) => look_up_or_insert!(payment_cache, payment_hash),
```

Comment:
This does another map lookup and then matches on the entry, just like here? I believe the macro can be dropped here and the macro removed after in-lining in the one remaining place its used.

Reply:
There needs to be at least one extra look-up here in order to get an `OccupiedEntry`, though it currently does two more it seems. We really want to call `hash_map::Entry::insert` here since it returns an `OccupiedEntry` and thus wouldn't need any extra look-ups. Unfortunately, that function is experimental. 🙁 

I updated this to insert directly and look-up once more only. Likewise for `pay_pubkey` in a later commit. Also, inlined the macro in `handle_event`.

Reply:
Can we refactor things in terms of `&mut ` instead of `OccupiedEntry` and do the drop based on return value?

Reply:
Yeah, though the drop would require another look-up. Probably ok for the error case, I suppose. Would also need to pass the `PaymentHash` around, too.